### PR TITLE
Add retries to Help proxy request

### PR DESCRIPTION
**Pull request by juliasilge**
Addresses https://github.com/posit-dev/positron/issues/3753

Folks in that issue ⤴️ are experiencing problems seeing R Help pages in the Help pane. Some things we can notice:

- None of us internally have been able to reproduce the problem.
- Some folks _do_ occasionally see the Help page they are querying for! Seems like an intermittent problem.
- We see errors in logs like "operation was canceled: connection closed before message completed".
- Some folks have their problems totally solved after installing a new version of Positron.

Given all that, seems like we may just need to retry the request, or at the least that is a first possible step as outlined by @DavisVaughan in https://github.com/posit-dev/positron/issues/3753#issuecomment-2245984690.

### CODE REVIEW

File: crates/ark/src/help_proxy.rs [hunk 1]
```diff
@@ -242,3 +253,22 @@ async fn preview_img(params: web::Query<PreviewRdParams>) -> HttpResponse {
...
+    let out = actix_web::http::header::HeaderValue::from_bytes(x.as_bytes());
+
+    // We've checked these are the same underlying structure, so assert that this works.
+    let mut out = out.unwrap();
+
+    // Set the one other field that defaults to `false`, just in case.
+    out.set_sensitive(x.is_sensitive());
+
+    out
+}
```
Comment (juliasilge) [inline crates/ark/src/help_proxy.rs:274]: Joint work with @DavisVaughan who discovered that:

- actix-web re-exports http 0.2.7's `HeaderValue`
- reqwest re-exports http 1.0.0's `HeaderValue`

Here is the commit where reqwest changed this:
https://github.com/seanmonstar/reqwest/commit/eb94f26919e881177ad3d9cd172f47d1a8263799

File: crates/ark/src/help_proxy.rs [hunk 2]
```diff
@@ -128,8 +132,14 @@ async fn proxy_request(req: HttpRequest, app_state: web::Data<AppState>) -> Http
     // Add query from original request back to URL.
     target_url.set_query(Some(query));
 
+    // Set up reqwest client with 3 retries.
+    let retry_policy = ExponentialBackoff::builder().build_with_max_retries(3);
+    let client = ClientBuilder::new(Client::new())
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .build();
+
```
Comment (juliasilge) [inline crates/ark/src/help_proxy.rs:140]: Shall we start with 3 retries?

File: crates/ark/src/help_proxy.rs [hunk 3]
```diff
@@ -17,6 +17,10 @@ use actix_web::HttpServer;
 use harp::exec::RFunction;
 use harp::exec::RFunctionExt;
 use mime_guess::from_path;
+use reqwest::Client;
+use reqwest_middleware::ClientBuilder;
+use reqwest_retry::policies::ExponentialBackoff;
+use reqwest_retry::RetryTransientMiddleware;
```
Comment (juliasilge) [inline crates/ark/src/help_proxy.rs:23]: The reqwest crate doesn't have any retry mechanism, but we can use these related crates to set up middleware and retries.

File: crates/ark/Cargo.toml [hunk 4]
```diff
@@ -41,7 +41,9 @@ notify = "6.0.0"
 once_cell = "1.17.1"
 parking_lot = "0.12.1"
 regex = "1.10.0"
-reqwest = { version = "0.11.20", features = ["json"] }
+reqwest = { version = "0.12.5", features = ["json"] }
+reqwest-retry = "0.6.1"
+reqwest-middleware = "0.3.3"
```
Comment (DavisVaughan) [inline crates/ark/Cargo.toml:46]: Good news is that both reqwest-retry and reqwest-middleware are smart enough to set `default-features = false` for their reqwest dependency
https://github.com/TrueLayer/reqwest-middleware/blob/6cd4eb93f8567bece4526f050cd6f7c2bde7371e/reqwest-middleware/Cargo.toml#L24
https://github.com/TrueLayer/reqwest-middleware/blob/6cd4eb93f8567bece4526f050cd6f7c2bde7371e/reqwest-retry/Cargo.toml#L19

Making #487 still viable even after this

File: crates/ark/src/help_proxy.rs [hunk 5]
```diff
@@ -128,8 +132,14 @@ async fn proxy_request(req: HttpRequest, app_state: web::Data<AppState>) -> Http
     // Add query from original request back to URL.
     target_url.set_query(Some(query));
 
+    // Set up reqwest client with 3 retries.
```
Comment (DavisVaughan) [inline crates/ark/src/help_proxy.rs:135]:
```suggestion
    // Set up reqwest client with 3 retries (posit-dev/positron#3753).
```

File: crates/ark/src/help_proxy.rs [hunk 6]
```diff
@@ -128,8 +132,14 @@ async fn proxy_request(req: HttpRequest, app_state: web::Data<AppState>) -> Http
...
+    let retry_policy = ExponentialBackoff::builder().build_with_max_retries(3);
```
Comment (DavisVaughan) [inline crates/ark/src/help_proxy.rs:136]: For the curious, here are the other default options that get set for the backoff:

https://docs.rs/retry-policies/0.4.0/src/retry_policies/policies/exponential_backoff.rs.html#148-151

Definitions
https://docs.rs/reqwest-retry/latest/reqwest_retry/policies/struct.ExponentialBackoff.html#fields

**Review by DavisVaughan — APPROVED** [review]
Seems right to me

**Pull request merged**